### DISCUSSION TITLE
There are some garbled characters in the displayed results

### DISCUSSION THREAD

**Discussion by tyf2018** [Taitava/obsidian-shellcommands#170] [Debugging and testing]
Hi， @Taitava 

SC version: 0.11.0
obsidian version: 0.13.26

I use the following command and it works almost fine:
```
Write-Host("Size: ", [math]::Round((Get-Item {{file_path:absolute}}).length/1KB), "kB Modified: ", (Get-Item {{file_path:absolute}}).LastWriteTime)
```

However, there are some garbled characters in the displayed results. I guess it is caused by not correctly identifying Chinese, because if I change "Modified" to Chinese, the garbled characters will also be displayed.

![image](https://user-images.githubusercontent.com/56342788/156090971-c7a74407-825e-421a-adc4-7ea82659f41b.png)

Replenish:
When there is an emoji expression in the path, an error will be reported.
![image](https://user-images.githubusercontent.com/56342788/156093341-40a2d778-0c7e-4a41-9d23-44e06482b3d4.png)

Replenish 2：
Is there a way to turn off this prompt? I can't find out what the error code is for it.
![image](https://user-images.githubusercontent.com/56342788/156098672-578e06dd-fdda-4ec3-9fa2-5afdbcfd570b.png)

**Comment by Taitava** [comment]
> Hi， @Taitava
> 
> SC version: 0.11.0 obsidian version: 0.13.26
> 
> I use the following command and it works almost fine:
> 
> ```
> Write-Host("Size: ", [math]::Round((Get-Item {{file_path:absolute}}).length/1KB), "kB Modified: ", (Get-Item {{file_path:absolute}}).LastWriteTime)
> ```
> 
> However, there are some garbled characters in the displayed results. I guess it is caused by not correctly identifying Chinese, because if I change "Modified" to Chinese, the garbled characters will also be displayed.

Yea, it's known (#157) but I don't know yet what would be a correct way to solve it. For a workaround, can you try this:

```PowerShell
$OutputEncoding = [console]::InputEncoding = [console]::OutputEncoding = New-Object System.Text.UTF8Encoding; Write-Host("Size: ", [math]::Round((Get-Item {{file_path:absolute}}).length/1KB), "kB Modified: ", (Get-Item {{file_path:absolute}}).LastWriteTime)
```

The `$OutputEncoding = [console]::InputEncoding = [console]::OutputEncoding = New-Object System.Text.UTF8Encoding; ` part tries to set both input and output encoding to `utf-8`. The corrupted characters are proabably due to an incorrect _output_ encoding, which should be fixed by this, although preceding a shell command with this kind of long workaround thing is not so nice to use.


> Replenish: When there is an emoji expression in the path, an error will be reported.

For this issue, I'd suspect it might relate to _input_ encoding, maybe? But you can try if the above workaround fixes this too. If it does not fix it, can you write here the name/path of the file, including the emoji too? 🙂 



> Replenish 2： Is there a way to turn off this prompt? I can't find out what the error code is for it. ![image](https://user-images.githubusercontent.com/56342788/156098672-578e06dd-fdda-4ec3-9fa2-5afdbcfd570b.png)

There's no way to turn it off at the moment. It doesn't have an error code as it happens in a stage where the shell command is not executed yet (happens during parsing the variables, and when variable parsing fails, the execution is cancelled without any error code).

In the future, there will be a way for defining default values for variables, which will allow the execution. This will probably happen the same time when #146 / #159 is implemented, which is among the next things on my radar, but it might take some time until they are done and released.

**Comment by tyf2018** [comment]
> $OutputEncoding = [console]::InputEncoding = [console]::OutputEncoding = New-Object System.Text.UTF8Encoding; Write-Host("Size: ", [math]::Round((Get-Item {{file_path:absolute}}).length/1KB), "kB Modified: ", (Get-Item {{file_path:absolute}}).LastWriteTime)

Thanks, this works fine.

> For this issue, I'd suspect it might relate to input encoding, maybe? But you can try if the above workaround fixes this too. If it does not fix it, can you write here the name/path of the file, including the emoji too? 🙂

As long as an emoji expression appears in the file name, an error will be reported. Something like this: 待办（看板）🐓 。Without 🐓, it works fine.

**Reply by Taitava** [reply]
> As long as an emoji expression appears in the file name, an error will be reported. Something like this: 待办（看板）🐓 。Without 🐓, it works fine.

Thank you. I think I was able to move one step forward with inspecting this, but I don't have a solution yet. First off, I was able to reproduce the same problem with the file name you mentioned, including the 🐓 emoji.

Second, I ran the following shell command in Windows Terminal (PowerShell 7). The idea was to check if PowerShell itself can handle the emoji in the file name, and yes, it can.
```powershell
$OutputEncoding = [console]::InputEncoding = [console]::OutputEncoding = New-Object System.Text.UTF8Encoding; Write-Host("Size: ", [math]::Round((Get-Item "待办（看板）🐓.md").length/1KB), "kB Modified: ", (Get-Item "待办（看板）🐓.md").LastWriteTime)
```

The result in Windows Terminal (PowerShell 7):
![kuva](https://user-images.githubusercontent.com/13002142/156373678-5b73f931-1c73-4b4d-a365-70280b0a4151.png)
The file size 0 kB is correct, I only typed just a short text to the file, far below 1 kilobyte.

And another run just in case in Windows Terminal, but PowerShell 5:
![kuva](https://user-images.githubusercontent.com/13002142/156373991-c562c790-a275-49f9-b78a-20b1a62e60a1.png)

The command works in PowerShell 5 too, although the file name looks like questionmarks in the command.

The results show that the issue is somewhere in the _Shell commands_ plugin.

I happened to find out something interesting in shell command preview:

![kuva](https://user-images.githubusercontent.com/13002142/156375241-88807101-0632-4292-b267-cb2b6d23dcb3.png)

The shell command does not do anything useful, it just echoes the file path plus the emoji. But the preview text is interesting: The emoji that I manually types into the end of the command, is displayed correctly in the gray preview text. Then again, the emoji that comes from the `{{file_path:relative}}` variable, is corrupted. I suspect the problem is somewhere in the variable parsing logic, something there does not support this kind of special characters.

I'll continue the investigation and try to come up with a bug fix regarding the emoji problem.

**Reply by Taitava** [reply]
#171 will fix the emoji bug.

**Reply by Taitava** [reply]
@tyf2018 can you check if the emoji bug is fixed? 🙂

**Reply by tyf2018** [reply]
> @tyf2018您可以检查表情符号错误是否已修复？ 🙂

Great, it's been fixed!

**Comment by Taitava** [comment]
# A solution for _{{file_path}}: No file is active at the moment..._

> > Replenish 2： Is there a way to turn off this prompt? I can't find out what the error code is for it. ![image](https://user-images.githubusercontent.com/56342788/156098672-578e06dd-fdda-4ec3-9fa2-5afdbcfd570b.png)
> 
> There's no way to turn it off at the moment. It doesn't have an error code as it happens in a stage where the shell command is not executed yet (happens during parsing the variables, and when variable parsing fails, the execution is cancelled without any error code).
> 
> In the future, there will be a way for defining default values for variables, which will allow the execution. This will probably happen the same time when #146 / #159 is implemented, which is among the next things on my radar, but it might take some time until they are done and released.

@tyf2018 SC version `0.12.0` (released today) now has these abilities:
 - [Define default values for variables](https://publish.obsidian.md/shellcommands/Variables/Default+values#Cancel+execution+without+displaying+error+messages).
 - [Cancel execution silently if a variable is not available](https://publish.obsidian.md/shellcommands/Variables/Default+values#Cancel+execution+without+displaying+error+messages). No errors displayed. This might be a good choice for the situation quoted above.
